File: Medium/874. Walking Robot Simulation.java
import java.util.HashSet;

class Solution {
    public int robotSim(int[] commands, int[][] obstacles) {
        final var obsSet = new HashSet<Integer>();
        for (var obs : obstacles) {
            obsSet.add(obs[0] + obs[1] * 600001);
        }
        final var dirs = new int[] { 0, 1, 0, -1, 0 };
        int currDir = 0, currX = 0, currY = 0;
        int res = 0;
        for (int com : commands) {
            if (com == -1) {
                currDir = (currDir + 1) % 4;
                continue;
            }
            if (com == -2) {
                currDir = (currDir + 3) % 4;
                continue;
            }
            for (int i = 0; i < com; i++) {
                int nextX = currX + dirs[currDir], nextY = currY + dirs[currDir + 1];
                if (obsSet.contains(nextX + nextY * 600001)) {
                    break;
                }
                currX = nextX;
                currY = nextY;
            }
            res = Math.max(res, currX * currX + currY * currY);
        }
        return res;
    }
}

File: Medium/874. Walking Robot Simulation.py
class Solution:
    def robotSim(self, commands, obstacles):
        blocked = set(map(tuple, obstacles))
        directions = ((0, 1), (1, 0), (0, -1), (-1, 0))
        x = y = d = 0
        max_dist = 0

        for cmd in commands:
            if cmd == -1:
                d = (d + 1) % 4  # turn right
            elif cmd == -2:
                d = (d + 3) % 4  # turn left
            else:
                while cmd > 0:
                    nx = x + directions[d][0]
                    ny = y + directions[d][1]

                    # check obstacle
                    if (nx, ny) in blocked:
                        break

                    x = nx
                    y = ny

                    max_dist = max(max_dist, x * x + y * y)
                    cmd -= 1

        return max_dist
